### DISCUSSION TITLE
RabbitMQ CRUD Queues/Exchanges?

### DISCUSSION THREAD

**Discussion by guttizin** [rabbitmq/rabbitmq-dotnet-client#1523] [General]
Am I doing/thinking wrong?
Why is there no basic functionality in RabbitMQ-Client-[dotnet|java] such as: list queues, list queue bindings?

Shouldn't the client support all the basic CRUD functionality for a queue/exchange?

I'm looking for functionality within IModel, which is where Create(QueueDeclare / ExchangeDeclare) and Delete(QueueDelete...) are, but where are the Retrieve and Update functionality from CRUD? Would it just be through the API? /api/queues and /api/exchanges?

**Comment by lukebakken** [comment]
> but where are the Retrieve and Update functionality from CRUD?

Updating a queue involves deleting it and re-creating it, which you can do from this library. This is a very rare thing to do, however.

By "retrieving" I assume you mean these operations:
* Retrieve a specific queue by name - use a passive declaration for that.
* Retrieve a list of all queues - use the HTTP API.

> Would it just be through the API? /api/queues and /api/exchanges?

Depending on what you want to do, yes. None of the RabbitMQ AMQP 0.9.1 client libraries support the operations that the HTTP API provides.

**Reply by guttizin** [reply]
I really appreciate all the support!

I'm doing some tests, and I would like to manage queues and exchanges programmatically.
I will use the API, as suggested.

At the end of the tests, with the scenario prepared, I will use `export_definitions` and `import_definitions` as michael suggested too.
Really thanks!!

**Comment by michaelklishin** [comment]
I'll take a wild guess that this has to do with seeding a large-ish topology. There's already a [library and programming language-neutral solution](https://www.rabbitmq.com/docs/definitions) that has been around for over a decade.

**Reply by guttizin** [reply]
Thank you for your support.

I'm doing some tests on the viability and performance of the system. Initially a test to find out how the system behaves with multiple queues, multiple exchanges and multiple bindings.

To do this, I simulated a scenario and would like to reset it (delete all queues, exchanges and bindings) and create a new one, programmatically.
But to reset, the functions require the name of the queues and exchange as a parameter, hence my question.
[remounting docker's image is also an alternative]

At the end of the tests, I would like to have a script to create the production scenario. **The alternative of using `import_definitions` will be perfect for this case. I didn't know about this yet.**

Thanks!!

**Reply by lukebakken** [reply]
Great guess @michaelklishin
